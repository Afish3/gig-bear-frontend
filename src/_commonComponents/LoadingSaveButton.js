import React, { useState, useRef, useEffect } from 'react';
import Box from '@mui/material/Box';
import CircularProgress from '@mui/material/CircularProgress';
import { green } from '@mui/material/colors';
import Button from '@mui/material/Button';

const LoadingSaveButton = ({ text, handleClick }) => {
  const [loading, setLoading] = useState(false);
  const [success, setSuccess] = useState(false);
  const timer = useRef();

  const buttonSx = {
    ...(success && {
      bgcolor: green[500],
      '&:hover': {
        bgcolor: green[700],
      },
    }),
  };

  useEffect(() => {
    return () => {
      clearTimeout(timer.current);
    };
  }, []);

  const handleButtonClick = () => {
    if (!loading) {
      setSuccess(false);
      setLoading(true);
      timer.current = window.setTimeout(async () => {
        await handleClick();
        setSuccess(true);
        setLoading(false);
      }, 1000);
    }
  };

  return (
    <Box sx={{ display: 'flex', alignItems: 'center' }}>
      <Box sx={{ m: 1, position: 'relative' }}>
        <Button
          variant="contained"
          sx={buttonSx}
          disabled={loading}
          onClick={handleButtonClick}
          size='large'
          color='success'
        >
          {text}
        </Button>
        {loading && (
          <CircularProgress
            size={24}
            sx={{
              color: green[500],
              position: 'absolute',
              top: '50%',
              left: '50%',
              marginTop: '-12px',
              marginLeft: '-12px',
            }}
            data-testid="loading-indicator"
          />
        )}
      </Box>
    </Box>
  );
}

export default LoadingSaveButton;