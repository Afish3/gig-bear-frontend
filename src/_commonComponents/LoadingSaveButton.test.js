import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import LoadingSaveButton from './LoadingSaveButton';

describe('LoadingSaveButton Component', () => {
  it('renders button with default state', () => {
    render(<LoadingSaveButton text="Save" handleClick={() => {}} />);
    const button = screen.getByRole('button', { name: 'Save' });
    expect(button).toBeInTheDocument();
    expect(button).not.toBeDisabled();
  });

  it('handles button click and triggers loading state', () => {
    const mockHandleClick = jest.fn();
    render(<LoadingSaveButton text="Save" handleClick={mockHandleClick} />);
    const button = screen.getByRole('button', { name: 'Save' });
    fireEvent.click(button);

    expect(button).toBeDisabled();
    expect(screen.getByTestId('loading-indicator')).toBeInTheDocument();
  });

  it('handles success state after a delay', async () => {
    const mockHandleClick = jest.fn(() => Promise.resolve());
    render(<LoadingSaveButton text="Save" handleClick={mockHandleClick} />);
    const button = screen.getByRole('button', { name: 'Save' });
    userEvent.click(button);

    await waitFor(() => {
      expect(mockHandleClick).toHaveBeenCalled();
    });

    const loadingIndicator = screen.queryByTestId('loading-indicator');
    expect(loadingIndicator).not.toBeInTheDocument();

    expect(button).not.toBeDisabled();

    const saveButton = screen.getByRole('button', { name: 'Save' });
    expect(saveButton).toHaveStyle({
      backgroundColor: 'rgb(76, 175, 80)',
    });
  });

  it('cleans up the timer on unmount', () => {
    const mockClearTimeout = jest.spyOn(global, 'clearTimeout');
    const { unmount } = render(<LoadingSaveButton text="Save" handleClick={() => {}} />);
    unmount();
    expect(mockClearTimeout).toHaveBeenCalled();
  });

  it('matches snapshot before', () => {
    const { asFragment } = render(<LoadingSaveButton text="Save" handleClick={() => {}} />);
    expect(asFragment).toMatchSnapshot();
  });

  it('matches snapshot after', async () => {
    const mockHandleClick = jest.fn(() => Promise.resolve());
    const { asFragment } = render(<LoadingSaveButton text="Save" handleClick={mockHandleClick} />);

    const button = screen.getByRole('button', { name: 'Save' });
    userEvent.click(button);

    await waitFor(() => {
      expect(mockHandleClick).toHaveBeenCalled();
    });

    expect(asFragment).toMatchSnapshot();
  });
});