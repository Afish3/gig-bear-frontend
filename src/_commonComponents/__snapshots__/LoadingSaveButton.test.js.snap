// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingSaveButton Component matches snapshot after 1`] = `[Function]`;

exports[`LoadingSaveButton Component matches snapshot before 1`] = `[Function]`;
